include/led.h

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _ZEPHYR_LED_H
+#define _ZEPHYR_LED_H
+
+/**
+ * @file
+ * @brief Public LED driver APIs
+ */
+
+#include <zephyr/types.h>
+#include <device.h>
+
+/**
+ * @typedef led_api_blink()
+ * @brief Callback API for blinking an LED
+ *
+ * @see led_blink() for argument descriptions.
+ */
+typedef int (*led_api_blink)(struct device *dev, u32_t led,
+			     u32_t delay_on, u32_t delay_off);
+
+/**
+ * @typedef led_api_set_brightness()
+ * @brief Callback API for setting brightness of an LED
+ *
+ * @see led_set_brightness() for argument descriptions.
+ */
+typedef int (*led_api_set_brightness)(struct device *dev, u32_t led,
+				      u16_t value);
+/**
+ * @typedef led_api_on()
+ * @brief Callback API for turning on an LED
+ *
+ * @see led_on() for argument descriptions.
+ */
+typedef int (*led_api_on)(struct device *dev, u32_t led);
+
+/**
+ * @typedef led_api_off()
+ * @brief Callback API for turning off an LED
+ *
+ * @see led_off() for argument descriptions.
+ */
+typedef int (*led_api_off)(struct device *dev, u32_t led);
+
+/**
+ * @brief LED driver API
+ *
+ * This is the mandatory API any LED driver needs to expose.
+ */
+struct led_driver_api {
+	led_api_blink blink;
+	led_api_set_brightness set_brightness;
+	led_api_on on;
+	led_api_off off;
+};
+
+/**
+ * @brief Blink an LED
+ *
+ * This routine starts blinking an LED forever with the given time period
+ *
+ * @param dev LED device
+ * @param led LED channel/pin
+ * @param delay_on Time period (in milliseconds) an LED should be ON
+ * @param delay_off Time period (in milliseconds) an LED should be OFF
+ * @return 0 on success, negative on error
+ */
+static inline int led_blink(struct device *dev, u32_t led,
+			    u32_t delay_on, u32_t delay_off)
+{
+	const struct led_driver_api *api = dev->driver_api;
+
+	return api->blink(dev, led, delay_on, delay_off);
+}
+
+/**
+ * @brief Set LED brightness
+ *
+ * This routine sets the brightness of a LED to the given value.
+ * Calling this function after led_blink() won't affect blinking.
+ *
+ * @param dev LED device
+ * @param led LED channel/pin
+ * @param value Brightness value to set
```

Comment:
What are the units for brightness?

Reply:
This would result in device-specific scales for brightness. I suggest we specify brightness as a fraction of full scale, represented in Q15 format.

Reply:
Okay. How do I implement it in API? Lets say that we want to pass `0.5` as the brightness value:

Reply:
I don't know of a general way to do this with a preprocessor macro. There are libraries to do it in runtime such as https://www.keil.com/pack/doc/CMSIS/DSP/html/group__float__to__x.html, but I wouldn't recommend using that for a constant. I suppose you could create some macros for common brightness values, such as 1.0, 0.5, 0.25.

Reply:
If we create macros for common values such as BRIGHT_1, BRIGHT_0_25 etc... then aren't we limiting the user in terms of granularity? What if the user wants to use 0.10 for some obvious reasons? Even the linux LED API is using the enum definitions for specifying the brightness values but I prefer to give the user full flexibility to control brightness. Then in that case a plain `uint` makes sense!

What do you think?

Reply:
I'm not suggesting that we use an enum in the API, because you're right, that would restrict the granularity unnecessarily. I'm suggesting that we normalize the brightness so that full scale (or max brightness) is represented by the same `u32_t` value across multiple drivers. A Q31 format would mean full scale (~1.0) is represented by the value 2^31-1=0x7FFF_FFFF. One-tenth brightness (0.1) would be represented as 2^31/10=0x0CCC_CCCC